packages/datadog-instrumentations/src/rhea.js

```diff
@@ -0,0 +1,215 @@
+
+addHook({ name: 'rhea', versions: ['>=1'], file: 'lib/link.js' }, obj => {
+  const startSendCh = channel('apm:rhea:send:start')
+  const startReceiveCh = channel('apm:rhea:receive:start')
+
+  const Sender = obj.Sender
+  const Receiver = obj.Receiver
+  shimmer.wrap(Sender.prototype, 'send', send => function (msg, tag, format) {
+    if (!canTrace(this)) {
+      // we can't handle disconnects or ending spans, so we can't safely instrument
+      return send.apply(this, arguments)
+    }
+
+    const name = getResourceNameFromSender(this)
+    const { host, port } = getHostAndPort(this.connection)
+
+    startSendCh.publish({ name, host, port, msg })
+    const delivery = send.apply(this, arguments)
+    const asyncResource = new AsyncResource('bound-anonymous-fn')
+    delivery[dd] = { asyncResource }
+    addToInFlightDeliveries(this.connection, delivery)
+    try {
+      return delivery
+    } catch (err) {
+      errorCh.publish(err)
+
+      throw err
+    } finally {
+      endCh.publish(undefined)
+    }
+  })
+
+  shimmer.wrap(Receiver.prototype, 'dispatch', dispatch => function (eventName, msgObj) {
+    if (!canTrace(this)) {
+      // we can't handle disconnects or ending spans, so we can't safely instrument
+      return dispatch.apply(this, arguments)
+    }
+
+    if (eventName === 'message' && msgObj) {
+      startReceiveCh.publish({ msgObj, connection: this.connection })
+
+      if (msgObj.delivery) {
+        const asyncResource = new AsyncResource('bound-anonymous-fn')
+        msgObj.delivery[dd] = { asyncResource }
+        msgObj.delivery.update = wrapDeliveryUpdate(msgObj.delivery, msgObj.delivery.update)
+        addToInFlightDeliveries(this.connection, msgObj.delivery)
+      }
+      try {
+        return dispatch.apply(this, arguments)
+      } catch (err) {
+        errorCh.publish(err)
+
+        throw err
+      } finally {
+        endCh.publish(undefined)
+      }
+    }
+
+    return dispatch.apply(this, arguments)
+  })
+  return obj
+})
+
+addHook({ name: 'rhea', versions: ['>=1'], file: 'lib/connection.js' }, Connection => {
+  shimmer.wrap(Connection.prototype, 'dispatch', dispatch => function (eventName, obj) {
+    if (eventName === 'disconnected') {
+      const error = obj.error || this.saved_error
+      if (this[inFlightDeliveries]) {
+        this[inFlightDeliveries].forEach(delivery => {
+          const { asyncResource } = delivery[dd]
+          asyncResource.runInAsyncScope(() => {
+            errorCh.publish(error)
+            finish(delivery, null)
+          })
+        })
+      }
+    }
+    return dispatch.apply(this, arguments)
+  })
+  return Connection
+})
+
+addHook({ name: 'rhea', versions: ['>=1'], file: 'lib/session.js' }, (Session) => {
+  patchCircularBuffer(Session.prototype, Session)
+  return Session
+})
+
+function canTrace (link) {
+  return link.connection && link.session && link.session.outgoing
+}
+
+function getResourceNameFromSender (sender) {
```

Comment:
A resource name is a tracing concept and shouldn't be present in the instrumentation.